ares__parse_into_addrinfo.c

```diff
@@ -0,0 +1,249 @@
+            status = ARES_EBADRESP;
+            break;
+          }  /* LCOV_EXCL_STOP */
+
+          node = ares__append_addrinfo_node(&head_node);
+          if (!node)
+            {
+              status = ARES_ENOMEM;
+              goto failed_stat;
+            }
+
+          sin = ares_malloc(sizeof(struct sockaddr_in));
+          if (!sin)
+            {
+              status = ARES_ENOMEM;
+              goto failed_stat;
+            }
+          memset(sin, 0, sizeof(struct sockaddr_in));
+          memcpy(&sin->sin_addr.s_addr, aptr, sizeof(struct in_addr));
+          sin->sin_family = AF_INET;
+
+          node->ai_addr = (struct sockaddr *)sin;
+          node->ai_family = AF_INET;
+          node->ai_addrlen = sizeof(struct sockaddr_in);
+
+          /* Ensure that each A TTL is no larger than the CNAME TTL. */
+          if (rr_ttl < ai->cname.ttl)
+            node->ai_ttl = rr_ttl;
+          else
+            node->ai_ttl = ai->cname.ttl;
+
+          status = ARES_SUCCESS;
+        }
+      else if (rr_class == C_IN && rr_type == T_AAAA
+          && rr_len == sizeof(struct ares_in6_addr)
+          && strcasecmp(rr_name, hostname) == 0)
+        {
+          got_aaaa = 1;
+          if (aptr + sizeof(struct ares_in6_addr) > abuf + alen)
+          {  /* LCOV_EXCL_START: already checked above */
+            ares_free(rr_name);
+            status = ARES_EBADRESP;
+            break;
+          }  /* LCOV_EXCL_STOP */
+
+          node = ares__append_addrinfo_node(&head_node);
+          if (!node)
+            {
+              status = ARES_ENOMEM;
+              goto failed_stat;
+            }
+
+          sin6 = ares_malloc(sizeof(struct sockaddr_in6));
+          if (!sin6)
+            {
+              status = ARES_ENOMEM;
+              goto failed_stat;
+            }
+
+          memset(sin6, 0, sizeof(struct sockaddr_in6));
+          memcpy(&sin6->sin6_addr.s6_addr, aptr, sizeof(struct ares_in6_addr));
+          sin6->sin6_family = AF_INET6;
+
+          node->ai_addr = (struct sockaddr *)sin6;
+          node->ai_family = AF_INET6;
+          node->ai_addrlen = sizeof(struct sockaddr_in6);
+
+          /* Ensure that each A TTL is no larger than the CNAME TTL. */
+          if (rr_ttl < ai->cname.ttl)
+            node->ai_ttl = rr_ttl;
+          else
+            node->ai_ttl = ai->cname.ttl;
+
+          status = ARES_SUCCESS;
+        }
+      else if (rr_class == C_IN && rr_type == T_CNAME)
+        {
+          got_cname = 1;
+          /* Decode the RR data and replace the hostname with it. */
+          status = ares__expand_name_for_response(aptr, abuf, alen, &rr_data,
+                                                  &len);
+          if (status != ARES_SUCCESS)
+            break;
+
+          ares_free(hostname);
+          hostname = rr_data;
+
+          /* Take the min of the TTLs we see in the CNAME chain. */
+          if (ai->cname.ttl > rr_ttl)
+            ai->cname.ttl = rr_ttl;
+
```

Comment:
no reason to record a minimum, instead record the ttl associated with the cname entry in the array